subprojects/base-services/src/main/java/org/gradle/util/internal/GUtil.java

```diff
@@ -571,4 +573,22 @@ public static boolean isSecureUrl(URI url) {
         return !"http".equalsIgnoreCase(scheme);
     }
 
+    /**
+     * Checks the entry name for zip-slip vulnerable sequences.
+     *
+     * @throws IllegalArgumentException if the entry contains vulnerable sequences
+     */
+    public static String safeZipEntryName(String name) {
+        if (isUnsafeZipEntryName(name)) {
+            throw new IllegalArgumentException(format("'%s' is not a safe zip entry name.", name));
+        }
+        return name;
+    }
+
+    public static boolean isUnsafeZipEntryName(String name) {
```

Comment:
Why do we use this, instead of the recommended validation, i.e., checking that the canonical path is within bounds?

From https://snyk.io/research/zip-slip-vulnerability
```java
  String canonicalDestinationDirPath = destinationDir.getCanonicalPath();
  File destinationfile = new File(destinationDir, e.getName());
  String canonicalDestinationFile = destinationfile.getCanonicalPath();
  if (!canonicalDestinationFile.startsWith(canonicalDestinationDirPath + File.separator)) {
    throw new ArchiverException("Entry is outside of the target dir: " + e.getName());
  }
```

Reply:
Because of how costly the recommended validation turned out to be when applied to every single entry. 

The good thing about the recommendation is that it is very general. Gradle can assume more about the zip entries.

Reply:
@leonard84, getting up-to-date performance numbers for the impact of the recommended validation in https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_Stage_ReadyforNightly_PerformanceTests_Trigger/48608023?showRootCauses=true, let's see.

Reply:
Makes sense. FTR:
```
Speed Results for test project 'archivePerformanceProject' with tasks visitZip: we're slower than 7.5-commit-1408963aacb with 99% confidence.
  |      Difference: 833 ms slower (833 ms), 42.98%
```